### PR TITLE
[MRG] Don't scatter data inside dask workers

### PR DESCRIPTION
**Pull request by pierreglaser**
Workaround for dask/distributed#3703, which affects the `joblib-dask` integration.
When scattering data inside `dask` worker, task using this data can often end up being cancelled, see the referenced issue for more information.

This situation typically happens inside `joblib` during nested `Parallel` calls (whenn the inner parallel call scatters some data).

As suggested by @mrocklin, another workaround is to use `hash=False` inside `client.scatter` calls -- I'd like to compare the two solutions via a few benchmarks before merging this PR.

Related, but does not fix entirely #959 

This also needs tests.

### CODE REVIEW

File: joblib/_dask.py [hunk 1]
```diff
@@ -43,6 +44,14 @@ def is_weakrefable(obj):
         return False
 
 
+def _in_dask_worker():
+    try:
+        worker = get_worker()
+    except ValueError:
+        worker = None
+    return worker
```
Comment (mrocklin) [inline joblib/_dask.py:52]: This is probably a little bit more robust.

```suggestion
    from distributed.utils import thread_state 
    return hasattr(thread_state, "execution_state")
```

**Comment by ogrisel** [comment]
I am also worried by sending large data chunks to the scheduler process.

Although I do not fully understand the impact `hash=True` and why we get naming collisions I think it would be worth implementing this solution.

Along side we might also want to implement an option to the joblib dask backend  to choose between global nested parallelism (by submitting nested calls to the scheduler as we currently do) and a local nested variant as suggested by @mrocklin in https://github.com/dask/distributed/issues/3703#issuecomment-619446911

Maybe the local variant should be active by default as it's likely to be lower over-head in 99% of the cases but can potentially lead to under-subscription of the cluster in some rare cases.

**Comment by mrocklin** [comment]
hash=False should work well in our case I think.  When you call
client.scatter within a worker all you do is put the array into the
workers's .data dictionary.  So this is free and simple.  The data is very
likely to only stay here and not be replicated, so I think that the
benefits of hashing for deduplication are low while the cost of hashing
every array on every task is likely to be somewhat high.  I suspect/hope
that not hashing resolves the problems that we're seeing and leads us to
near optimal behavior.

**Comment by ogrisel** [comment]
It worked!

@pierreglaser can you please update the changelog?

**Comment by pierreglaser** [comment]
Sure.

File: joblib/test/test_dask.py [hunk 2]
```diff
@@ -251,6 +251,38 @@ def test_auto_scatter(loop):
...
+
+    NUM_INNER_TASKS = 10
+    NUM_OUTER_TASKS = 10
+
+    def my_sum(x, i, j):
+        # print(f"running inner task {j} of outer task {i}")
+        return np.sum(x)
+
+    def outer_function_joblib(array, i):
+        # print(f"running outer task {i}")
```
Comment (pierreglaser) [inline joblib/test/test_dask.py:267]: leftover debug print call

File: joblib/_dask.py [hunk 3]
```diff
@@ -269,15 +269,22 @@ async def maybe_to_futures(args):
...
+                    if f is None:
                         if is_weakrefable(arg) and sizeof(arg) > 1e3:
                             # Automatically scatter large objects to some of
                             # the workers to avoid duplicated data transfers.
                             # Rely on automated inter-worker data stealing if
                             # more workers need to reuse this data
                             # concurrently.
+                            # set hash=False - nested scatter calls (i.e
+                            # calling client.scatter inside a dask worker)
+                            # using hash=True often raises CancelledError,
```
Comment (pierreglaser) [inline joblib/_dask.py:282]: typo:
```suggestion
                            # using hash=True often raise CancelledError,
```

File: joblib/_dask.py [hunk 4]
```diff
@@ -269,15 +269,22 @@ async def maybe_to_futures(args):
                     try:
                         f = call_data_futures[arg]
                     except KeyError:
+                        pass
```
Comment (pierreglaser) [inline joblib/_dask.py:272]: Now, we won't get confusing `KeyErrors` if something wrong happens during `scattering` :)

File: joblib/test/test_dask.py [hunk 5]
```diff
@@ -251,6 +251,38 @@ def test_auto_scatter(loop):
...
+@pytest.mark.parametrize("retry_no", list(range(2)))
+def test_nested_scatter(loop, retry_no):
+
+    np = pytest.importorskip('numpy')
+
+    NUM_INNER_TASKS = 10
+    NUM_OUTER_TASKS = 10
+
+    def my_sum(x, i, j):
+        # print(f"running inner task {j} of outer task {i}")
```
Comment (pierreglaser) [inline joblib/test/test_dask.py:263]: leftover debug print call

**Comment by ogrisel** [comment]
Merged! Thank you very much @pierreglaser and @mrocklin.

**Comment by mrocklin** [comment]
Great.  I am glad to see this.  I'm also now curious what performance is like now as a result of this.  Are there other issues that we should resolve?

**Pull request merged**